api/api.go

```diff
@@ -94,3 +117,46 @@ func (api *API) Update(cfg *config.Config, resolveTimeout time.Duration) error {
 
 	return api.v2.Update(cfg, resolveTimeout)
 }
+
+// makeLimiter returns an HTTP middleware that sets a timeout for HTTP requests
+// and also limits the number of concurrently processed GET requests to the
+// given number.
+//
+// If timeout is < 1, no timeout is enforced.
+//
+// If concurrency is < 1, GOMAXPROCS is used as the concurrency limit but at least 8.
+//
+// The returned middleware serves http.StatusServiceUnavailable (503) for requests that
+// would exceed the number.
+func makeLimiter(timeout time.Duration, concurrency int) func(http.Handler) http.Handler {
+	if concurrency < 1 {
+		concurrency = runtime.GOMAXPROCS(0)
+		if concurrency < 8 {
+			concurrency = 8
+		}
+	}
+	inFlightSem := make(chan struct{}, concurrency)
+
+	return func(h http.Handler) http.Handler {
+		concLimiter := http.HandlerFunc(func(rsp http.ResponseWriter, req *http.Request) {
+			if req.Method == http.MethodGet { // Only limit concurrency of GETs.
+				select {
+				case inFlightSem <- struct{}{}: // All good, carry on.
+					defer func() { <-inFlightSem }()
+				default:
+					http.Error(rsp, fmt.Sprintf(
```

Comment:
I guess logging this would defeat the purpose of a rate limiter trying to prevent overload, but how about exposing a counter for this?

Reply:
Oh sure. There is even a middleware in client_golang just for that.

Reply:
I have added a "requests in flight" gauge.

In different news: The api/v2 calls are currently not instrumented at all because they act on the Handler level while the current instrumentation acts on the Router level only. I guess that's an oversight. Fixing it is a bit more involved. I suggest to do it after the refactoring you have already started. I left a TODO note.

Reply:
Thanks for adding the middleware. Yes, API v2 is not yet instrumented. On my agenda.

Reply:
Slight update: For some reason (which I cannot remember), Stuart and I decided to have the in-flight middleware not partitioned by method. Since the limit only applies to GET requests, I have now only instrumented the GET requests, not even using the middleware (which in turn now includes the api/v2 calls into this particular instrumentation).

I'm not sure if we should instrument the in-flight requests in general, including method and handler labels. We can still estimate the overlapping request load from the latency and rate. Thoughts welcome. If it is in general useful to instrument in-flight requests partitioned by method, we should change the middleware in client_golang.

Reply:
I _think_ the reason for not partitioning the `InstrumentHandlerInFlight` middleware is that you cannot partition by status code as that one is only determined once the request is not in flight anymore. We could, of course, still partition by method. However, the middleware is already in popular use. If we really want to break all those users (it's an easy code change to adjust to a GaugeVec without labels, but still), it should be worth it.

I'd say for now, another v0.10 thing. Let's go with the GET only instrumentation here for now... (not actually using the middleware, in fact).

Reply:
I guess there is a lot of cleanup in terms of our instrumentation, which we should probably do in a follow up patch.

Never the less we could also do "alertmanager_http_requests_in_flight { method="GET" }`, what do you think?

Reply:
I think the `method` label only makes sense if we have plan to instrument the in-flight gauge for other methods, too. It kind of suggests that by being a label and not part of the metric name.
If you think we should eventually instrument all methods that way, we can certainly go down the label approach.

Reply:
I will add the label (lower-case `get` for now, as this is (sadly) the norm in client_golang, cf. https://github.com/prometheus/client_golang/issues/387 ) plus a TODO to instrument other method's, too.

I'll also add a Registerer as one of the options.